### ISSUE TITLE
API: Register Electron App as Protocol Handler

### ISSUE THREAD

**Issue by feross** [atom/electron#4857]
There's no simple, cross-platform way to register protocol handlers in Electron. Every OS requires a different solution. This seems like something that Electron should smooth over.

There's already a W3C API for this: [`Navigator.registerProtocolHandler`](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/registerProtocolHandler). Can Electron implement it?

**Comment by felixrieseberg** [comment]
I'd be in favor of that and happy to help with code :+1:

**Comment by feross** [comment]
For OS X, this seems to require modifying the `Info.plist`, unless there's some other mechanism to set this at runtime that I haven't discovered yet.

For Windows, this is actually pretty straightforward after lots of research. I implemented it in a couple dozen lines of code, without even requiring a native dependency. https://github.com/feross/webtorrent-app/blob/2b555e12b868b0610838e55f98a0b1d6705f49d6/main/register-protocol-handler.js#L7-L46 This uses [`winreg`](https://www.npmjs.com/package/winreg) to modify the Windows registry. It just calls out to the `REG` command line program with `child_process`, so it's pure JavaScript.

**Comment by felixrieseberg** [comment]
Winreg has a few issues for things like protocol handlers and file handlers across versions of Windows, which is why my team at Microsoft[ built this (native) alternative](https://github.com/CatalystCode/windows-registry-node), ensuring the correct ACLs for the operation.

**Comment by feross** [comment]
What potential ACL issues exist with my approach? Isn't `HKEY_CURRENT_USER\Software\Classes` always writable by the current user?

**Comment by felixrieseberg** [comment]
It is, but certain environments might not have access to regedit. Our solution just does it directly in c++, removing any dependencies.

This is mostly interesting for Visual Studio Code, Atom, and other apps that have *a lot* of users, bringing edge cases with them.

**Comment by feross** [comment]
@felixrieseberg Okay, fair enough. Thanks for explaining the potential issue. Right now, our app doesn't have any native dependencies and we're trying to keep it that way as long as possible so it's simpler to package.

For larger apps, or for a built-in Electron API, the C++ approach sounds right.

**Comment by paulcbetts** [comment]
> It is, but certain environments might not have access to regedit.

This is correct, many environments inexplicably use Group Policy to block REG.exe but not editing the registry in general. Yes, this _is_ dumb.

**Comment by paulcbetts** [comment]
Actually, I'm full of it, @deepak1556 tracked down https://code.google.com/p/chromium/codesearch#chromium/src/chrome/browser/shell_integration.h&q=shell_integration&sq=package:chromium&l=45 which looks to have implementations for all platforms, I'm a :+1: on this

**Comment by felixrieseberg** [comment]
Nice, good find! Quick question though for you, @paulcbetts - the `navigator` API wasn't really meant for this and I'm concerned that we're just replacing an existing API with a slightly different function. I'd rather make a new API, maybe somewhere on `app`. Thoughts?

**Comment by paulcbetts** [comment]
@felixrieseberg Yeah, I'd implement the same API on `app` - it's really a per-app Thing, and if in the future Chromium implemented this API, we'd conflict if we tried to patch it in

**Comment by felixrieseberg** [comment]
Cool. I'll take a look at this, we'll need this in Ghost, too.

**Comment by feross** [comment]
> if in the future Chromium implemented this API, we'd conflict if we tried to patch it in

Chrome already implements the API :) You can see it in action here: https://instant.io (Look at the icon in the right of the Location bar)

**Comment by deepak1556** [comment]
> Chrome already implements the API :) You can see it in action here: https://instant.io (Look at the icon in the right of the Location bar)

`navigator.registerProtocolHandler` handles protocol within the User Agent, it is similar to setting up a redirect request job with Electrons' protocol module. Can we edit the title as its misleading to the feature requested here ?

**Comment by felixrieseberg** [comment]
Update: OS X is done, but Windows is a bit tricky (depending on how wide our range of supported versions is).

**Comment by feross** [comment]
@felixrieseberg Nice work! I see you used the code comment from WebTorrent inline :)

For Linux, feel free to re-use what we've written for WebTorrent.

.desktop file: https://github.com/feross/webtorrent-app/blob/73d5a4e1ab47c7139df836f4b7755f332d00a9c4/static/webtorrent.desktop
function: https://github.com/feross/webtorrent-app/blob/73d5a4e1ab47c7139df836f4b7755f332d00a9c4/main/register-handlers.js#L16-L31

**Comment by felixrieseberg** [comment]
Given that I know literally _nothing_ about how to do this on Linux, it'd probably be better if you or someone else who knows what they're doing in Linux desktop environments would tackle this in a separate PR (once #4896 is merged). You'd just have to hook in [here](https://github.com/atom/electron/pull/4896/files#diff-c23b2418b0e2c377a43e7c77b07d4e94R40).

**Comment by feross** [comment]
I also know next-to-nothing about Linux. I hacked this together for our app. Perhaps @grunjol could pitch in? He's sent a bunch of great Linux PRs to our app.

**Comment by grunjol** [comment]
I am not expert neither but made some hacks

In the past linux DE's make whatever they want, but some years ago they follow XDG specifications.

The app can be installed for the local user or for the system, in case of the system i recommend to create a distribution package (deb, rpm, etc.), the following steps are for users apps.

As @feross said, the first thing is create the [.desktop file](https://specifications.freedesktop.org/desktop-entry-spec/latest/) for the application (ie myapp.desktop). In the file you have a MIME field you can add the list of protocols the app can handle.
The .desktop file can be installed using the _xdg-desktop-menu_ command (the package xdg-utils should be installed) or you can do it manually copying it to $HOME/.local/share/applications/
You can copy an icon (ie: myapp.png) file to  $HOME/.local/share/icons/
That basically adds the app to your menus.

Adding the MIME line in the .desktop file doesn't automatically add the handler, you need to associate it adding the line 

`x-scheme-handler/myapp=myapp.desktop;` 

to either $HOME/.config/mimeapps.list or $HOME/.local/share/applications/mimeapps.list (check the existence of one or the other in that order)

You can check if its working using the command line 

`xdg-mime query default x-scheme-handler/myapp`

or test it invoking

 `xdg-open myapp://<example.protocol.data>`

For more information about this:
https://wiki.archlinux.org/index.php/default_applications
https://wiki.debian.org/MIME

Maybe usefull how chromium implements shell integration https://src.chromium.org/viewvc/chrome/trunk/src/chrome/browser/shell_integration_linux.cc

**Comment by bbondy** [comment]
FWIW, I implemented the real navigator.registerProtocolHandler which gives websites the ability to request registration of a protocol here: https://github.com/brave/electron/commit/3a2f05bb03b1a76421dd93b344183c9a8bffbef1

Good examples of this is gmail now prompts to set as default handler for mailto: and coinbase for bitcoin: urls.

Unfortunately it requires KeyedService though which we only have in our branch with extensions support.  I'd be happy to upstream one day if Electron's stance changes on not taking in extensions support or if they get the needed support for KeyedService otherwise.

**Comment by caryfitzhugh** [comment]
@felixrieseberg I'm looking into using this stuff and can't figure out the details of how you actually catch the event when the OS triggers your protocol.

In Mac I can catch it with 
```
app.on('open-url', ....
```

But the docs say that is mac only.

How are you actually using this ?  :)

**Comment by caryfitzhugh** [comment]
@felixrieseberg I think I found that it's just executing the CLI - I was hoping for something more like DDE - where it can trigger an event if the process is already started.

https://msdn.microsoft.com/en-us/library/aa767914(v=vs.85).aspx

https://msdn.microsoft.com/en-us/library/windows/desktop/hh127429(v=vs.85).aspx

If you have any insights, please let me know. :)

**Comment by caryfitzhugh** [comment]
and - I've discovered this gem:

https://github.com/electron/electron/blob/master/docs/api/app.md#appmakesingleinstancecallback

Already in electron, and looks like it will fwd my command line args to the first instance. Perfect.

**Comment by felixrieseberg** [comment]
@caryfitzhugh Seems like you already figured out how to use it ;-)

**Comment by jamgold** [comment]
@caryfitzhugh how do I communicate with the already running app?

**Comment by caryfitzhugh** [comment]
Could you explain your issue a bit more and I can try to help...
 communicate as in... ipc?


On Mar 15, 2017 6:13 PM, "Jan Hendrik Mangold" <notifications@github.com>
wrote:

@caryfitzhugh <https://github.com/caryfitzhugh> how do I communicate with
the already running app?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/electron/electron/issues/4857#issuecomment-286896158>,
or mute the thread

**Comment by jamgold** [comment]
@caryfitzhugh thanks for responding so quickly. I initiate the app / mainWindow in main.js and have my logic in render.js where I again access app over remote object. How can I trigger an even on the app?

**Comment by jamgold** [comment]
I figured out how to accomplish this. In main.js I either call ```mainWindow.webContents.reload()``` which does the trick, or I use ```mainWindow.webContents.send( 'ping','wohoo' )``` and add the following to render.js
```
require('electron').ipcRenderer.on('ping', function(event, message) {
    console.log(message);  // Prints "whoooooooh!"
});
```

**Comment by caryfitzhugh** [comment]
Wonderul. Sorry I couldn't get back to you quicker. Away from my computer
at the moment...  glad you figured it out :)

**Comment by Gozala** [comment]
This issue is closed but from what I can tell `navigator.registerProtocolHandler` still does not work. From what I can tell `navigator.registerProtocolHandler` succeeds (as long origin of the registered URL has same origin as document performing registration). But then attempt to load registered URL triggers [`openExternal`](https://electronjs.org/docs/api/session#sessetpermissionrequesthandlerhandler) permission handler. Accepting / Rejecting it seems to make no difference.

I think this issue mentioned `navigator.registerProtocolHandler` but was really about [`app.setAsDefaultProtocolClient`](https://electronjs.org/docs/api/app#appsetasdefaultprotocolclientprotocol-path-args) instead.
